streams/src/main/java/org/apache/kafka/streams/processor/internals/assignment/DefaultStateConstrainedBalancedAssignor.java

```diff
@@ -0,0 +1,252 @@
+
+    /**
+     * This assignment algorithm guarantees that all task for which caught-up clients exist are assigned to one of the
+     * caught-up clients. Tasks for which no caught-up client exist are assigned best-effort to satisfy the balance
+     * factor. There is not guarantee that the balance factor is satisfied.
+     *
+     * @param statefulTasksToRankedClients ranked clients map
+     * @param balanceFactor balance factor (at least 1)
+     * @return assignment
+     */
+    @Override
+    public Map<ID, List<TaskId>> assign(final SortedMap<TaskId, SortedSet<ClientIdAndLag<ID>>> statefulTasksToRankedClients,
+                                        final int balanceFactor) {
+        final Map<ID, List<TaskId>> assignment = initAssignment(statefulTasksToRankedClients);
+        final Map<TaskId, List<ID>> tasksToCaughtUpClients = tasksToCaughtUpClients(statefulTasksToRankedClients);
+        assignTasksWithCaughtUpClients(assignment, tasksToCaughtUpClients, statefulTasksToRankedClients);
+        assignTasksWithoutCaughtUpClients(assignment, tasksToCaughtUpClients, statefulTasksToRankedClients);
+        balance(assignment, balanceFactor, statefulTasksToRankedClients, tasksToCaughtUpClients);
+        return assignment;
+    }
+
+    /**
+     * Initialises the assignment with the clients in the ranked clients map and assigns to each client an empty list.
+     *
+     * @param statefulTasksToRankedClients ranked clients map
+     * @return initialised assignment with empty lists
+     */
+    private Map<ID, List<TaskId>> initAssignment(final SortedMap<TaskId, SortedSet<ClientIdAndLag<ID>>> statefulTasksToRankedClients) {
+        final Map<ID, List<TaskId>> assignment = new HashMap<>();
+        final SortedSet<ClientIdAndLag<ID>> clientIdAndLags = statefulTasksToRankedClients.values().stream()
+            .findFirst().orElseThrow(() -> new IllegalStateException("The list of clients and lags must not be empty"));
+        for (final ClientIdAndLag<ID> clientIdAndLag : clientIdAndLags) {
+            assignment.put(clientIdAndLag.clientId(), new ArrayList<>());
+        }
+        if (assignment.isEmpty()) {
+            throw new IllegalStateException("Clients within an assignment must not be empty");
+        }
+        return assignment;
+    }
+
+    /**
+     * Maps tasks to clients with caught-up states for the task.
+     *
+     * @param statefulTasksToRankedClients ranked clients map
+     * @return map from tasks with caught-up clients to the list of client candidates
+     */
+    private Map<TaskId, List<ID>> tasksToCaughtUpClients(final SortedMap<TaskId, SortedSet<ClientIdAndLag<ID>>> statefulTasksToRankedClients) {
+        final Map<TaskId, List<ID>> taskToCaughtUpClients = new HashMap<>();
+        for (final SortedMap.Entry<TaskId, SortedSet<ClientIdAndLag<ID>>> taskToRankedClient : statefulTasksToRankedClients.entrySet()) {
```

Comment:
```suggestion
        for (final Map.Entry<TaskId, SortedSet<ClientIdAndLag<ID>>> taskToRankedClient : statefulTasksToRankedClients.entrySet()) {
```

Just to resolve a warning about referencing the subclass instead of the interface.

Reply:
How did you get that warning? How can this produce a warning?

Reply:
It's a warning you can enable in IntelliJ IDEA. Not a huge deal.